hs_tools_resource/models.py

```diff
@@ -230,9 +234,56 @@ def remove(cls, element_id):
 class ToolIcon(AbstractMetaDataElement):
     term = 'ToolIcon'
     value = models.CharField(max_length=1024, blank=True, default="")
+    data_url = models.TextField(blank=True, default="")
+
+    @classmethod
+    def _validate_tool_icon(cls, url):
+        try:
+            response = requests.get(url, verify=False)
+        except:
+            raise ValidationError("Failed to read data from given url.")
+        if response.status_code != 200:
+            raise ValidationError("Failed to read data from given url.")
+        image_size_mb = float(response.headers["content-length"])/1024.0/1024.0
+        if image_size_mb > 1:
+            raise ValidationError("Icon image size should be less than 1MB.")
+        image_type = imghdr.what(None, h=response.content)
+        if image_type not in ["png", "gif", "jpeg"]:
+            raise ValidationError("Supported icon image types are png, gif and jpeg")
+        base64_string = base64.b64encode(response.content)
+        data_url = "data:image/{image_type};base64,{base64_string}".\
+                   format(image_type=image_type, base64_string=base64_string)
+        return data_url
+
+    @classmethod
+    def create(cls, **kwargs):
+        if 'value' in kwargs:
+            url = kwargs["value"]
+            data_url = cls._validate_tool_icon(url)
+
+            metadata_obj = kwargs['content_object']
+            new_meta_instance = ToolIcon.objects.create(content_object=metadata_obj)
+            new_meta_instance.value = url
+            new_meta_instance.data_url = data_url
+            new_meta_instance.save()
```

Comment:
@hyi 
Thanks for this comment. I think I see your point, but I am a little bit confused about difference between a field and a property. Is a property an independent data stored in db? or is it just a dynamic representation of an existing field?
Our current workflow is user inputs/changes a icon url on UI or through rest,
1) backend saves the url in field 'value' 
2) backend retrieves icon data from that url, encodes data as a base64 string and saves it in field "data_url".
Then, UI can read icon data (the base64 string) from our db instead of dynamically reading an external icon url, which is the whole point of this PR.  I am concerning if I change data_url field to a property, the base64 string will be generated on the fly each time UI reads it.